File: src/main/scala/MnMcount.scala
package main.scala

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._


object MnMcount {
    def main(args: Array[String]): Unit = {
        val spark = SparkSession
            .builder
            .appName("MnMcount")
            .master("local[*]")
            .getOrCreate()

        if (args.length < 2) {
          println("Usage: MnMcount <mnm_file_dataset> <output_data_path")
          sys.exit(1)
        }

        val mnmFile = args(0)
        val outputPath = args(1)

        val schema = StructType(Array(
            StructField("State", StringType, false),
            StructField("Color", StringType, false),
            StructField("Count", IntegerType, false),
            ))
        val mnmDF = spark.read
            .schema(schema)
            .format("csv")
            .option("header", "true")
            .load(mnmFile)
        val caCountMnNDF = mnmDF
            .select("State", "Color", "Count")
            .where(col("State") === "CA")
            .groupBy("State", "Color")
            .sum("Count")
            .withColumnRenamed("sum(Count)", "TotalCount")
            .orderBy(desc("TotalCount"))
        val caCountMnNDF_double = caCountMnNDF.withColumn("DoubleTotalCount", col("TotalCount") * 2)

        caCountMnNDF.show(10)
        caCountMnNDF_double.show(10)

        caCountMnNDF.write.mode("overwrite").format("csv").save(outputPath)

        spark.stop()
    }
}

File: src/mnm_count.py
import argparse
from pyspark.sql import SparkSession
from pyspark.sql.types import *
from pyspark.sql.functions import *


def main(input_path, output_path):
    spark = (SparkSession
             .builder
             .master("local[*]")
             .appName("count mnm data with spark")
             .getOrCreate())

    schema = StructType([
        StructField("State", StringType(), False),
        StructField("Color", StringType(), False),
        StructField("Count", IntegerType(), False)
    ])
    mnm_df = (spark.read
              .schema(schema)
              .format("csv")
              .option("header", "true")
              .load(input_path))
    agg_mnm_df = (mnm_df
                  .select("State", "Color", "Count")
                  .where(col("State") == "CA")
                  .groupBy("State", "Color")
                  .sum("Count")
                  .withColumnRenamed("sum(Count)", "TotalCount")
                  .orderBy("TotalCount", ascending=True))
    agg_mnm_df.show(5, False)

    agg_mnm_df.write.mode("overwrite").format("csv").save(output_path)

    spark.stop()


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--input_path", required=True)
    parser.add_argument("--output_path", required=True)
    args = parser.parse_args()

    input_path =args.input_path
    output_path = args.output_path

    main(input_path, output_path)
